---
title: "BC1_LD.LV.eR"
author: "FLC"
date: '2022-05-07'
output: html_document
---
```{r}
library(ggplot2)
library(gridExtra)
library(reshape2)
library(mixOmics)
library(RColorBrewer)
library(VennDiagram)
library(DESeq2)
library(ggrepel)
library(stringr)

library(edgeR)  # What will make the magic happen

library(readODS)
#install.packages("readODS")
library(tidyr)
```


# **Metadata** loading
```{r load_metadata}
# Loading of 'Animals.csv'
Animals <- read.delim("/home/fllobet/Documents/RNA-Seq_TFM/Animals.csv", sep =";", row.names = 3)

# Addition of the 'X' to the row names of Aniamls to  match col names of the
# counts matrix
rownames(Animals) <- paste("X", rownames(Animals), sep = "")

# Selection of 'BC1_LD'
Animals.BC1_LD <- Animals[which(Animals[,2] == "BC1_LD"),]

# Selection of 'liver'
Animals.BC1_LD.LV <- Animals.BC1_LD[which(Animals.BC1_LD[,3] == "Liver"),]
```

# **Counts** loading

The lodad counts are the *integer counts* extracted from the corresponding _DESeq2 script_ where they were genearted by means of the `tximport` package.
```{r load_counts}
cnt.BC1_LD.LV <- read.delim("/home/fllobet/Documents/RNA-Seq_TFM/BC1_LD.LV/IntegerCOUNTS.BC1_LD.liver.csv", sep = ";", row.names = 1)

# Conversion into a matrix
cnt.BC1_LD.LV <- as.matrix(cnt.BC1_LD.LV)
```

# Removal of the 2 males
```{r male_remove}
# Removal of the 2 males from the metadata
Animals.BC1_LD.LV <- Animals.BC1_LD.LV[which(Animals.BC1_LD.LV[,4] != "M"),]

# Removal of the 2 males from the counts matrix
cnt.BC1_LD.LV <- cnt.BC1_LD.LV[, rownames(Animals.BC1_LD.LV)]
```


# Creation of a `DGEList` **object**
Like in `desEq2` where a `dds` object had to be created here a `DGEList` object must be created.
```{r DGEList_object}
compare <- relevel(as.factor(Animals.BC1_LD.LV$Group), ref = "L")

################ CHECKING COLS count matrix == ROWS metadata ###################
# The while loop serves to see the OK prompt when the orders has been modified. 
while (TRUE) {
  if (levels(as.factor(rownames(Animals.BC1_LD.LV) == 
                       colnames(cnt.BC1_LD.LV))) == TRUE) {
    # If all the rows == all the columns (only TRUE): 
    print("The columns of the counts matrix are the same as 
          the rows of the metadata")
    break # Exit infinite loop. 
} else {
  # Change the order of the columns of the count matrix to match the metadata
  # which is the order of the "COMPARE" vector. 
  cnt.BC1_LD.LV <- cnt.BC1_LD.LV[,rownames(Animals.BC1_LD.LV)]
}
}

dge.BC1_LD.LV <- DGEList(cnt.BC1_LD.LV, group = compare)
dge.BC1_LD.LV
```



# Filtering
The native `edgeR` function `filterByExpr()`. 
This function filters features that have **CPM above _k_ in _n_ samples** where:
- _n_ is the smallest group size (in our case both groups have same size)
- _k_ is the CPM equal to 10 / _L_ where _L_ is the **minimum library size**

In human words it will remove any feature that has more than 10 counts _(but normalizing by library depth)_ in at least > samples than the small group.
```{r filtering}
a <- dim(dge.BC1_LD.LV$counts)[1]

# Filtering
keep <- filterByExpr(dge.BC1_LD.LV)
dge.BC1_LD.LV <- dge.BC1_LD.LV[keep, , keep.lib.sizes=FALSE]

# Percentage of remove features
round(((dim(dge.BC1_LD.LV$counts)[1] - a) / a), 4)

# Re-computation of the library sizes
dge.BC1_LD.LV$samples$lib.size <- colSums(dge.BC1_LD.LV$counts)
```
# Normalization - Computation of **Normalization Factors**
The **normalization factors** take into account **library size** via the *TMM* method. These factors will be used later in the `edgeR`model, the used values within the model are analogous to **CPM**.

The Normalization Factors are **considered automatically** into the FC and p.value calculations. 
```{r normalization_factors}
# Demonstration of the variability in library size between samples
st <- expression(paste("Note the ", bold("heterogeneity"), " and hence the ", bold("importance of normalization")))

DF <- cbind(rownames(dge.BC1_LD.LV$samples), dge.BC1_LD.LV$samples)
colnames(DF) <- c("Sample", colnames(DF[2:ncol(DF)]))

# Saving the object 'DF' to be used for later creation of lib sizes plots for 
# the thesis
saveRDS(DF, file = "/home/fllobet/Documents/RNA-Seq_TFM/Extra_figures/eR_LIB/BC1_LD.LV.rds")

ggplot(DF[order(Animals$Group),], aes(x= Sample, y = lib.size*1e-6, fill = group)) + geom_bar(stat = "identity") + labs(title = "Libray size accross samples", subtitle = st, x = "Sample", y = "Library size (millions)") + scale_x_discrete(breaks = rownames(dge.BC1_LD.LV$samples[order(Animals.BC1_LD.LV$Group),]), labels = str_replace(str_replace(rownames(dge.BC1_LD.LV$samples[order(Animals.BC1_LD.LV$Group),]), "X", ""),"_LV", "")) + 
scale_fill_manual(name = "Group", values = c("#bdb76b","#2f847c")) + theme_bw()

# Computation of normalization factors (accounting for library size)
dge.BC1_LD.LV$samples <- calcNormFactors(dge.BC1_LD.LV)
dge.BC1_LD.LV$samples

# Readdition of the group data beause it gets removed in the previous step
dge.BC1_LD.LV$samples$group <- compare

################## CHECKING COLS DGE object == ROWS metadata ###################
# The while loop serves to see the OK prompt when the orders has been modified. 
while (TRUE) {
  if (levels(as.factor(rownames(Animals.BC1_LD.LV) == 
                       colnames(dge.BC1_LD.LV))) == TRUE) {
    # If all the rows == all the columns (only TRUE): 
    print("The columns of the counts matrix are the same as 
          the rows of the metadata")
    break # Exit infinite loop. 
} else {
  # Change the order of the columns of the count matrix to match the metadata
  # which is the order of the "COMPARE" vector. 
  cnt.BC1_LD.LV <- cnt.BC1_LD.LV[,rownames(Animals.BC1_LD.LV)]
}
}

```
# Disersion estimation
Following the advice of DCP, we use **only `common` dispersion** so the significant genes will be those that with **extremely large counts in some libraries (samples) and not all the libraries (samples)**.
```{r common_dispersion}
dge.BC1_LD.LV <- estimateCommonDisp(dge.BC1_LD.LV)

# Value of the common dispersion:
dge.BC1_LD.LV$common.dispersion
```

# Exploratory data analysis

## Generation of CPM
```{r CPM}
CPM.BC1_LD.LV <- cpm(dge.BC1_LD.LV)
```

## Generation of **_pseudoCPM_**
```{r eda_01_pseudocounts}
pseudoCnt.BC1_LD.LV <- log2(CPM.BC1_LD.LV + 1)
head(pseudoCnt.BC1_LD.LV)
```

## Count distribution example
```{r eda_03_cnt_distro}
# Generation of a data frame containing 'Counts' and 'pseudoCounts' of sample 1 
df <- data.frame(counts = dge.BC1_LD.LV$counts[ ,1], 
                 pseudo = pseudoCnt.BC1_LD.LV[ ,1])

p <- ggplot(data=df, aes(x = counts, y = ..density..)) +
  geom_histogram(color = "#4d5d53", fill = "#ace1af") + theme_bw() +
  ggtitle(paste0("Count distribution '", 
                 Animals.BC1_LD.LV$Animal[1], "_LV'")) +
  xlab("Counts") + ylab("")

p2 <- ggplot(data=df, aes(x = pseudo, y = ..density..)) +
  geom_histogram(color = "#4d5d53", fill = "#2f847c") + theme_bw() +
  ggtitle(paste0("Pseudocount distribution '", 
                 Animals.BC1_LD.LV$Animal[1], "_LV'")) +
  xlab(expression(log[2](counts + 1)))

grid.arrange(p, p2, ncol = 2)


```
## ALl samples count distribution
```{r all_sample_counts}
# Merge of the metadata with the pseudocounts data
df_all_pseudoCnt <- cbind(Animals.BC1_LD.LV, t(pseudoCnt.BC1_LD.LV))

# Melting to have a 'ggplot2' ready dataframe containg metada + pseudocounts
df_all_pseudoCnt <- melt(df_all_pseudoCnt, id.vars = c("Animal", "Backcross", "Tissue", "Sex", "Group", "Father", "Mother", "Paper"))
df_all_pseudoCnt$Animal <- as.character(df_all_pseudoCnt$Animal)


p <- ggplot(data = df_all_pseudoCnt, aes(x = Animal, y = value, fill = Group)) + 
  geom_boxplot() + ggtitle("count distributions") + 
  scale_fill_manual(labels = c("H", "L"), values = c("#556b2f", "#bdb76b")) +
  xlab("sample") + ylab("counts") + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90))
p
```

## MA-Plot: Reproducibility between samples
```{r eda_05_MAplot}
# Manual MA-Plot
df_ma <- data.frame("A" = (pseudoCnt.BC1_LD.LV[,1] + pseudoCnt.BC1_LD.LV[, 2])/2,
                    "M" = pseudoCnt.BC1_LD.LV[,1] - pseudoCnt.BC1_LD.LV[,2])
p <- ggplot(data = df_ma, aes(x = A, y = M)) + geom_point(alpha = 0.2, color = "#4d5d53") + theme_bw()
p
```
## Overdispersion
```{r eda_06_overdispersion}

df_ovd <- data.frame(mean = apply(cnt.BC1_LD.LV[, Animals.BC1_LD.LV$Group == "L"], 1, mean),
                     var = apply(cnt.BC1_LD.LV[, Animals.BC1_LD.LV$Group == "L"], 1, var))

df_ovd <- df_ovd[df_ovd$mean <= 5000, ]

p_ovd <- ggplot(df_ovd, aes(x = mean, y = var)) + geom_point(color = "#ace1af") + 
  theme_bw() + geom_abline(aes(intercept=0, slope=1)) + ggtitle("Variance versus means in counts of LOW group") + ylab("Variance")
p_ovd
```
For most genes the variance is higher than the mean, that is the common **overdispersionm** of RNA-Seq data. 

## Heatmap
```{r eda_07_euclidHeatmap}
# Creation of a matrix of the euclidean distances of the pseudo-counts data. 
df_hm <- as.matrix(dist(t(pseudoCnt.BC1_LD.LV)))
# Colors of the heatmap. 
cim_color <- colorRampPalette(rev(brewer.pal(9, "YlGnBu")))(16)

cim(df_hm, color = cim_color, symkey = FALSE)

```

## PCA
```{r eda_08_pca}

# Selection of the top 500 SD features
# SD for all the rows
rowSD <- apply(pseudoCnt.BC1_LD.LV, 1, sd)
# Ordering by SD
pseudoCnt.BC1_LD.LV.top500 <- pseudoCnt.BC1_LD.LV[order(rowSD,
                                                        decreasing = TRUE),]
# Selection of the top 500 more variable features
pseudoCnt.BC1_LD.LV.top500 <- pseudoCnt.BC1_LD.LV.top500[1:500,]
################################################################################
# Computation of the PCA
resPCA <- prcomp(pseudoCnt.BC1_LD.LV)
summary(resPCA)
resPCA$sdev
# Percentages of variation explained per PC
pov <- resPCA$sdev^2/sum(resPCA$sdev^2)
# Rounding in percentages:
pov <- (round(pov, digits = 4) * 100)

################################################################################
# Conserving the Principal domponents data (eigenvalues or eigenvectors?)
resPCA_pc12 <- as.data.frame(resPCA$rotation)
# Addition of the sample metadata
resPCA_pc12 <- cbind(resPCA_pc12, Animals.BC1_LD.LV)

rlg <- expression(paste(bold("rlog"), " transformation applied"))
capt <- expression(paste("Counts of ", bold("liver"), " samples of ", bold("BC1_LD")))

PCA.plot <- ggplot(resPCA_pc12, aes(x = PC1, y = PC2, color = Group)) + 
  geom_point(size=3) + 
  geom_text_repel(aes(label = str_replace(
    str_replace(rownames(Animals.BC1_LD.LV), "_LV", ""),"X", ""))) +
  xlab(paste0("PC1: ",pov[1],"% variance")) +
  ylab(paste0("PC2: ",pov[2],"% variance")) + 
  scale_color_manual(values = c("#2f847c", "#bdb76b")) +  
  labs(title = "PCA of the top 500 more variable features",  
       subtitle = rlg, caption = capt) + theme_bw()

PCA.plot
plotMDS(pseudoCnt.BC1_LD.LV.top500, col = c(rgb(47/225,132/225,124/225),
                                            rgb(189/225,183/225,107/225),
                                            rgb(189/225,183/225,107/225),
                                            rgb(47/225,132/225,124/225),
                                            rgb(189/225,183/225,107/225),
                                            rgb(47/225,132/225,124/225),
                                            rgb(189/225,183/225,107/225),
                                            rgb(47/225,132/225,124/225)), labels = str_replace(str_replace(colnames(pseudoCnt.BC1_LD.LV.top500), "X", ""),"_LV", ""))
legend("bottomright", legend = c("H", "L"), col = c(rgb(189/225,183/225,107/225),rgb(47/225,132/225,124/225)), pch = 19)
title(main = "MDS plot (PCA) of the top 500 more variable features", sub = "Using root-mean-square average of the top log2FC between samples (leading log2FC).", cex.sub = 0.8)
```
# Differential expression analysis: Pairwise exact test
The test is quite similar to the Fisher's exact test
```{r pairwise_exact_test}
# Lodaing of the file where I have the list of significant features from 
# 'Ramayo-Caldas et al. 2012'
yuli.BC1_LD.LV <- read_ods("/home/fllobet/Documents/RNA-Seq_TFM/BC1_LD.LV/DEG_Yuliaxis_2012.ods")

# Exact test (core statistical test)
res_BC1_LD.LV <- exactTest(dge.BC1_LD.LV)
# Vislualitzation of the top 10 most significant features
topTags(res_BC1_LD.LV)

################################################################################
# FDR Correction: Benjamini-Honchberg.
res_BC1_LD.LV$table <- 
  cbind(res_BC1_LD.LV$table, 
        p.adjust(res_BC1_LD.LV$table$PValue, method = "BH"))

# Confirmation of the performed comparison
res_BC1_LD.LV$comparison

################################################################################
# Storing of the results into a CSV file
res_BC1_LD.LV$table <- cbind(rownames(res_BC1_LD.LV$table), res_BC1_LD.LV$table)
res_BC1_LD.LV$table <- cbind(rownames(res_BC1_LD.LV$table), res_BC1_LD.LV$table)
res_BC1_LD.LV$table[,1:2] <- str_split_fixed(rownames(res_BC1_LD.LV$table),"_", n = 2) 

colnames(res_BC1_LD.LV$table) <- c("ENSEMBL ID", "Gene name", colnames(res_BC1_LD.LV$table)[3:6])

# BUSCARV over the lists - ENSEMBL ID

IDs_in_yuli <- merge(res_BC1_LD.LV$table, yuli.BC1_LD.LV, by="ENSEMBL ID")$`ENSEMBL ID`

# Printing of the number of mathcing features
length(IDs_in_yuli)

# Addition of this info in the output CSV of the significant genes
L <- dim(res_BC1_LD.LV$table)[2]
for (i in 1:dim(res_BC1_LD.LV$table)[1]) {
  if (res_BC1_LD.LV$table[i,1] %in% IDs_in_yuli) {
    res_BC1_LD.LV$table[i,L+1] <- "in Ramayo-Caldas et al. 2012"
    }
}
################################################################################
# BUSCARV over the lists - GENE NAME


GNs_in_yuli <- merge(res_BC1_LD.LV$table, yuli.BC1_LD.LV, by="Gene name")$`Gene name`

L <- dim(res_BC1_LD.LV$table)[2]
for (i in 1:dim(res_BC1_LD.LV$table)[1]) {
  if (res_BC1_LD.LV$table[i, 2] %in% GNs_in_yuli) {
    res_BC1_LD.LV$table[i, L + 1] <- "in Ramayo-Caldas et al. 2012"
  }
}

################################################################################
# Storing the results into a CSV file 
colnames(res_BC1_LD.LV$table) <- c(colnames(res_BC1_LD.LV$table)[1:5], "PADJValue", "Paper ID", "Paper Gene")
write.table(res_BC1_LD.LV$table, "/home/fllobet/Documents/RNA-Seq_TFM/BC1_LD.LV/edgeR/BC1_LD.LV_eResults.csv", sep = ";", row.names = FALSE)
```

```{r signficant_features}
# Return of significant features with a given alpha threshold:
signif_BC1_LD.LV <- decideTestsDGE(res_BC1_LD.LV, adjust.method = "BH", p.value = 0.05)

# Selection of the significant features
sel_signif_BC1_LD.LV <- which(signif_BC1_LD.LV[,1] != 0)

# Datafame of the Results + P.adjusted values of only the significant features
signif_BC1_LD.LV <- cbind(res_BC1_LD.LV$table[sel_signif_BC1_LD.LV,],
                          signif_BC1_LD.LV[sel_signif_BC1_LD.LV, 1])

################################################################################
# Printing of:
# 'n' of logFC > 0 
dim(signif_BC1_LD.LV[which(signif_BC1_LD.LV$logFC > 0 & signif_BC1_LD.LV$PADJValue < 0.05),])[1]
# 'n' of logFC < 0
dim(signif_BC1_LD.LV[which(signif_BC1_LD.LV$logFC < 0 & signif_BC1_LD.LV$PADJValue < 0.05),])[1]

# Storing the signficiant features into a CSV file
write.table(signif_BC1_LD.LV, "/home/fllobet/Documents/RNA-Seq_TFM/BC1_LD.LV/edgeR/BC1_LD.LV_eResults_Signif.05.csv", sep = ";", row.names = FALSE)
```

# ABS Fold-Change > 1.2
```{r fc_1.2}
fc12 <- log2(1.2)

# Filtering 
signif_BC1_LD.LV_FC_1.2 <- signif_BC1_LD.LV[which(signif_BC1_LD.LV$logFC > fc12 | signif_BC1_LD.LV$logFC < -fc12),]
dim(signif_BC1_LD.LV_FC_1.2)[1]
```
# ABS Fold-Change > 1.5
```{r fc_1.5}
fc15<- log2(1.5)

# Filtering 
signif_BC1_LD.LV_FC_1.5 <- signif_BC1_LD.LV[which(signif_BC1_LD.LV$logFC > fc15 | signif_BC1_LD.LV$logFC < -fc15),]
dim(signif_BC1_LD.LV_FC_1.5)[1]
```
```{r}
rlg.dge.BC1_LD.LV.cnts <- rlog(dge.BC1_LD.LV$counts)

# Selection of the top 500 SD features
# SD for all the rows
rowSD <- apply(rlg.dge.BC1_LD.LV.cnts, 1, sd)
# Ordering by SD
rlg.dge.BC1_LD.LV.cnts.top500 <- rlg.dge.BC1_LD.LV.cnts[order(rowSD,
                                                        decreasing = TRUE),]
# Selection of the top 500 more variable features
pseudoCnt.rlog_BC1_LD.LV.top500 <- rlg.dge.BC1_LD.LV.cnts.top500[1:500,]
################################################################################
# Computation of the PCA
resPCA <- prcomp(pseudoCnt.rlog_BC1_LD.LV.top500)
summary(resPCA)
resPCA$sdev
# Percentages of variation explained per PC
pov <- resPCA$sdev^2/sum(resPCA$sdev^2)
# Rounding in percentages:
pov <- (round(pov, digits = 4) * 100)

################################################################################
# Conserving the Principal domponents data (eigenvalues or eigenvectors?)
resPCA_pc12 <- as.data.frame(resPCA$rotation)
# Addition of the sample metadata
resPCA_pc12 <- cbind(resPCA_pc12, Animals.BC1_LD.LV)

rlg <- expression(paste(bold("pseudocounts"), " (log2(counts + 1)) transformation applied"))
capt <- expression(paste("Counts of ", bold("liver"), " samples of ", bold("BC1_LD")))

PCA.plot <- ggplot(resPCA_pc12, aes(x = PC1, y = PC2, color = Group)) + 
  geom_point(size=3) + 
  geom_text_repel(aes(label = str_replace(
    str_replace(rownames(Animals.BC1_LD.LV), "_LV", ""),"X", ""))) +
  xlab(paste0("PC1: ",pov[1],"% variance")) +
  ylab(paste0("PC2: ",pov[2],"% variance")) + 
  scale_color_manual(values = c("#2f847c", "#bdb76b")) +  
  labs(title = "PCA of the top 500 more variable features",  
       subtitle = rlg, caption = capt) + theme_bw()

PCA.plot
```


# Number of reads
Inspired from Ramayo-Caldas et al. 2012 I want to take a look at the 
```{r}
colSums(cnt.BC1_LD.LV) / 10e6

26133508 / 10e6
23976233 / 10e6

```


# Check if the results match the results from Yuliaxis et al.



```{r check_yuli.ALL}
# Division of the Feature column in ENSEMBL ID + Gene Name



IDs_2022 <- rownames(cnt.BC1_LD.LV)
IDs_2022 <- cbind(rownames(cnt.BC1_LD.LV), IDs_2022)
IDs_2022[,1:2] <- str_split_fixed(rownames(cnt.BC1_LD.LV),"_", n = 2) 

colnames(IDs_2022) <- c("ENSEMBL ID", "Gene name")

# BUSCARV over the lists
IDs_in_yuli.ALl_ens <- merge(IDs_2022, yuli.BC1_LD.LV, by = "ENSEMBL ID")
dim(IDs_in_yuli.ALl_ens)[1]

IDs_in_yuli.ALL_gn <- merge(IDs_2022, yuli.BC1_LD.LV, by="Gene name")
dim(IDs_in_yuli.ALL_gn)[1]

IDs_in_yuli.CHNG_gn_ens <- merge(IDs_in_yuli.ALL_gn, IDs_in_yuli.ALl_ens, by.x = "Gene name", by.y = "Gene name.x")

# Addition of this data (ID & gene name) into 2 new columns:

# ENSEMBL ID
L <- dim(yuli.BC1_LD.LV)[2]
for (i in 1:dim(yuli.BC1_LD.LV)[1]) {
  if (yuli.BC1_LD.LV[i,1] %in% IDs_in_yuli.ALl_ens$`ENSEMBL ID`) {
    yuli.BC1_LD.LV[i,L+1] <- "SAME ID 2022"
    }
}

# Gene name
L <- dim(yuli.BC1_LD.LV)[2]
for (i in 1:dim(yuli.BC1_LD.LV)[1]) {
  if (yuli.BC1_LD.LV[i,3] %in% IDs_in_yuli.ALL_gn$`Gene name`) {
    yuli.BC1_LD.LV[i,L+1] <- "SAME GeneName 2022"
    }
}

write.table(yuli.BC1_LD.LV, "/home/fllobet/Documents/RNA-Seq_TFM/BC1_LD.LV/edgeR/yuli.BC1_LD.LV_2022.csv", sep = ";", row.names = FALSE)
```

```{r}
# Merging the 2 columns of 'Paper ID' and 'Paper Name' to paint the plot
res_BC1_LD.LV$table <- res_BC1_LD.LV$table %>% mutate(`Paper ID` = coalesce(`Paper ID`,`Paper Gene`)) %>%
         select(colnames(res_BC1_LD.LV$table)[1:7]) 

length(na.omit(res_BC1_LD.LV$table$`Paper ID`))

# Same plot but adapted for the master thesis document
BC1_LD.pv.Yuli <- ggplot(res_BC1_LD.LV$table[which(res_BC1_LD.LV$table$PADJValue < 1),], 
       aes(x = PADJValue, fill =  `Paper ID`)) + 
  geom_histogram(color = "#4d5d53") + 
  geom_vline(xintercept = 0.051, linetype='dashed', col = '#2f847c', size = 1.5) +
  geom_text(aes(x = 0.055, y = 85), label = "0.05", col = '#2f847c', size = 3.5, nudge_x = 0.05) +
  scale_fill_manual(name = "", values = c("#ace1af"), na.value = "#bdb76b", 
                    labels = c("in Yuli et al. 2012", 
                               "NOT in Yuli et al. 2012")) + 
  scale_y_continuous(expand = c(0,0), limits = c(0, 95), breaks = seq(0,95,20)) +
  scale_x_continuous(expand = c(0.005,0.005)) +
  labs(title = "LV:",
       x = "Adjusted p-values (FDR) distribution", y = "Frequency") + theme_bw() + 
  theme(text = element_text(family = "ubuntu", face = "bold"),
        legend.position = "none")

saveRDS(BC1_LD.pv.Yuli, 
        file = "~/Documents/RNA-Seq_TFM/Extra_figures/BC1_LV_vs_publ/BC1_LD.LV_vs_Yuli.rds")
```

