---
title: "BC1_LD.Liver.R"
author: "FLC"
date: '2022-03-31'
output:
  pdf_document: default
  html_document: default
---

```{r library_loading, include=FALSE}
#BiocManager::install("GenomicFeatures")
library(tximport) # To import the counts data from RSEM.
library(DESeq2)
library(vsn) # To generate the mean vs variance plots
library(pheatmap) # To include heatmaps
library(RColorBrewer)  # To automatically generate beautiful palettes.
library(ggplot2) # To make wonderful plots
library(randomForest) # To generate the MDSplot
library(GenomicFeatures)  # To open the GTF file used in the Mapping
library(ggrepel)  # To see who is who in the plots
library(stringr)  # To do fancy things with strings
library(snpStats)  # To generate the Q-Q Plot of the p-values
# BiocManager::install("snpStats")
```

```{r metadata_loading, echo = FALSE}
### METADATA of ALL SAMPLES
Animals <- read.delim("/home/fllobet/Documents/RNA-Seq_TFM/Animals.csv", sep = ";", stringsAsFactors = FALSE)

## Cleaning of the metadata
# Relocation of the Animal number as row names 
row.names(Animals) <- paste0("X",Animals[,3])
# The 'X' is necessary because R cannot use column names staring wiht a number.

### Subset of the METADATA from the BC1_LD liver samples:

# BC1_LD
Animals.BC1_LD <- Animals[which(Animals[,2] == "BC1_LD"),]
# Liver 
Animals.BC1_LD.LV <- Animals.BC1_LD[which(Animals.BC1_LD[,4] == "Liver"),] 
Animals.BC1_LD.LV

################################################################################
# Removal of unutilized objects
rm(Animals, Animals.BC1_LD)
```


## `RSEM` counts loading using `tximport`
```{r tximport_RSEM_counts_loading}
### IMPORT of counts using 'tximportData'
# Object that contains the path to all the '.genes.results' RSEM output files. 
BC1_LD.PORK_02 <- file.path("/home/fllobet/Documents/RNA-Seq_TFM/COUNTS/Genes/PK02", paste0(Animals.BC1_LD.LV$SAMPLE_NAME, ".genes.results"))
# Names of the columns
names(BC1_LD.PORK_02) <- paste0("X", Animals.BC1_LD.LV$SAMPLE_NAME) 
# The 'X' is necessary because R cannot use column names staring wiht a number.

### IMPORT
tx.BC1_LD.LV <- tximport(BC1_LD.PORK_02, type ="rsem", txIn = FALSE, txOut = FALSE)

# Header of the imported counts
head(tx.BC1_LD.LV$counts)
# Note that the number are still decimal.

################################################################################
# Removal of unutilized objects
rm(BC1_LD.PORK_02)
```

## Generation of the dds object

Before the RSEM-counts matrix can be converted into a `dds` object it is necessary to apply the **trick** proposed by the creator of `DESeq2` and `tximport`. This **trick** solves the 0-length features error. 
```{r trick, echo = FALSE}
# Removal of 0-length features and 0-counts features as instructed in:
# https://support.bioconductor.org/p/84304/
# The person who states this solution is the developer of 'tximport'. 
tx.BC1_LD.LV$length[tx.BC1_LD.LV$length == 0] <- 1

```
After the trick is done. The counts can effectively be converted into a `dds` object. 
```{r dds_object}
### Standard 'dds' object:
dds.BC1_LD.LV <- DESeqDataSetFromTximport(tx.BC1_LD.LV,
                              colData = Animals.BC1_LD.LV,
                              design = ~ Group) # This design is temporary. 

# Header of the counts now converted into postive integers: 
head(counts(dds.BC1_LD.LV)) # This counts are unormalized.

# Storing the positive-integer counts into a count matrix
CMX_dds.BC1_LD.LV <- counts(dds.BC1_LD.LV)
colnames(CMX_dds.BC1_LD.LV) <- str_replace(colnames(CMX_dds.BC1_LD.LV),"X","")
CMX_dds.BC1_LD.LV <- cbind(rownames(CMX_dds.BC1_LD.LV), CMX_dds.BC1_LD.LV) 
colnames(CMX_dds.BC1_LD.LV) <- c("Feature", colnames(CMX_dds.BC1_LD.LV)[2:dim(CMX_dds.BC1_LD.LV)[2]])
write.table(CMX_dds.BC1_LD.LV, file = "/home/fllobet/Documents/RNA-Seq_TFM/BC1_LD.LV/IntegerCOUNTS.BC1_LD.liver.csv", sep =";", row.names = FALSE)
################################################################################
# Removal of unutilized objects
rm(tx.BC1_LD.LV)
```
Note that now the **counts are positive integers instead** of decimals. This is important because the algorithms of `DESeq2` use the premise that the input counts data follow a **binomial negative distribution** which is **not continuous**.  

## Pre-filtering
The filter applied, instructed by JMF:
- minimum nº of counts per feature (row) == nº samples
```{r pre_filtering}
# Number of samples
N <- nrow(Animals.BC1_LD.LV)
# Features to keep
keep <- rowSums(counts(dds.BC1_LD.LV)) >= N

dim(dds.BC1_LD.LV)
# Application of the pre-filtering over the dds object. 
dds.BC1_LD.LV <- dds.BC1_LD.LV[keep,]

################################################################################
# Removal of unutilized objects
#rm(keep)
```

## Exploratory data analysis (EDA)
This step is key because it serves two main proposes:
- **Familiarization** with the data and its particularities.
- Identification of **technical artifacts**.

### Normalized counts
Any EDA of RNA-Seq data must never be done with raw counts as differences between  samples (library depth) might lead to wrong observations. 

`DESeq2` can perform **2 diffrent normalizations**:
- By **library size**
- By **library size** + **gene length**

By deafult `DESeq2` performs the first one as the authors say it is enough most of the times. 

```{r normalized_counts}

# Generation of normalization factors, DESeq2 on its own will choose if it 
# creates 'sj' or 'sij' 
dds.BC1_LD.LV <- estimateSizeFactors(dds.BC1_LD.LV)

# DESeq2 can use 2 Normalization constants: sj & sij
# 'sj' 
head(sizeFactors(dds.BC1_LD.LV)) 
# It is empty because gene-specific normalization factors. Why? I belive it is 
# linked to the fact we use RSEM for the counts.

# 'sij' 
head(normalizationFactors(dds.BC1_LD.LV))

### Normalization using above sij within the means (muij) of each gene-sample.
dds.BC1_LD.LV.norm.c <- counts(dds.BC1_LD.LV, normalized = TRUE) # This parameter is what normalized the count data using the above sij factors. 
print("Normalized counts")
head(dds.BC1_LD.LV.norm.c)
```
I believe that our data contains just `normalizationFactors` because the counts were obtained with RSEM which takes into account **feature length**. 

### Pseudocounts
```{r pseudocounts_generation}
dds.BC1_LD.LV.pseudo <- normTransform(dds.BC1_LD.LV)
```

### rlog transformation
Because CNAG and JVH both performed an rlog transformation I should also do it. After it, it is expected that the variance of small mean samples (the most problematic in  RNA-Seq experiments) get reduced. 
```{r}
vsd <- vst(dds.BC1_LD.LV, blind=FALSE)
# rlog transformation
rlg.dds.BC1_LD.LV <- rlog(dds.BC1_LD.LV, blind=FALSE)
```

Now that we have both pseudocounts and `rlog` transformed counts we can determine the effect of the rlog transformation
```{r}
# Mean vs SD plot of the pseudocounts
meanSdPlot(assay(dds.BC1_LD.LV.pseudo))

# Mean vs SD plot of the rlog transformed counts
meanSdPlot(assay(rlg.dds.BC1_LD.LV))

# Vsd tranfsormation appears to be worse for low mean features. Furthermore, in 
# general the variance is lower in the rlog transformation. 
# meanSdPlot(assay(vsd))
```

We can observe that effectively the **`rlog` transformation** causes that the **variance of low-mean features is greatly reduced**. Also a general reduction of the variance can be observed.

### Heatmaps
It is important to perform heat-maps to be able to interpret visually the counts matrix, as our human brain is not powerful enough to interpret a large numerical matrix.
```{r heatmaps}
# Selection of the top 20 most globally expressed features (rows).
select <- order(rowMeans(dds.BC1_LD.LV.norm.c),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds.BC1_LD.LV)[,c("Sex","Group")])

# Heat map of the normalized pseudocounts
pheatmap(assay(dds.BC1_LD.LV.pseudo)[select,], cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=FALSE, annotation_col=df)

# Heat map of the rlog transformed counts
pheatmap(assay(rlg.dds.BC1_LD.LV)[select,], cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=FALSE, annotation_col=df)

################################################################################
# Removal of unutilized objects
rm(df)
```
We see that the Albumin gene *ALB* is the **most expressed gene** in all the samples. This is likely being this protein known for its high abundance in the liver. It also looks like the **`rlog` transformation** made the colors palette more blue, meaning that the values got reduced probably spacing better the data. 

### Heatmap of euclidean distances
A heat map of this distance matrix gives us an overview over similarities and dissimilarities between samples. It might **help us to identify technical outliers**. 

```{r euclidean_distances_heatmap}
# Pseudocounts
# Computation of the counts Eculidean distances
pseudo_dist <- dist(t(assay(dds.BC1_LD.LV.pseudo)))
# Conversion into a matrix
pseudo_d.Mtx <- as.matrix(pseudo_dist)

# Addition of the SEX & GROUP information
rownames(pseudo_d.Mtx) <- paste(dds.BC1_LD.LV.pseudo$Sex, dds.BC1_LD.LV.pseudo$Group, sep = " ")
colnames(pseudo_d.Mtx) <- NULL

colors <- colorRampPalette(rev(brewer.pal(9, "YlGnBu")))(255)

# Heatmap
pheatmap(pseudo_d.Mtx,
         clustering_distance_rows=pseudo_dist,
         clustering_distance_cols=pseudo_dist,
         col=colors)

# rlog transformed counts
# Computation of the counts Eculidean distances
rlg_dist <- dist(t(assay(rlg.dds.BC1_LD.LV)))
# Conversion into a matrix
rlg_d.Mtx <- as.matrix(rlg_dist)

# Addition of the SEX & GROUP information
rownames(rlg_d.Mtx) <- paste(rlg.dds.BC1_LD.LV$Sex, rlg.dds.BC1_LD.LV$Group, sep = " ")
colnames(rlg_d.Mtx) <- rlg.dds.BC1_LD.LV$SAMPLE_NAME

colors <- colorRampPalette(rev(brewer.pal(9, "YlGnBu")))(255)

# Heatmap
pheatmap(rlg_d.Mtx,
         clustering_distance_rows=rlg_dist,
         clustering_distance_cols=rlg_dist,
         col=colors)

################################################################################
# Removal of unutilized objects
rm(pseudo_d.Mtx, pseduo_dist, pseudo_dist, rlg_dist, rlg_d.Mtx, colors)
```
It looks like sample 70154_LV differs a lot from some of the other samples but it does not have any importance.

### PCA
Related to the distance matrix is the PCA plot, which shows the samples in the 2D plane spanned by their first two principal components. This type of plot is useful for visualizing the overall effect of experimental covariates and batch effects. We might see separation according to Group or not. Furthmore this plot is what **JMF wants**

First I generate a PCA using all the features:
```{r whole_pca}
# By default plotPCA uses just the top 500 more variable features.

# Sutbtitle
rlg <- expression(paste(bold("rlog"), " transformation applied"))
capt <- expression(paste("Counts of ", bold("liver"), " samples of ", bold("BC1_LD")))

pcaData <- plotPCA(rlg.dds.BC1_LD.LV, intgroup=c("Sex", "Group"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=Group, shape=Sex)) +
  geom_point(size=3) + geom_text_repel(aes(label = str_replace(str_replace(rownames(pcaData), "_LV", ""),"X", ""))) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  scale_color_manual(values = c("#2f847c", "#bdb76b")) + labs(title = "PCA of the top 500 more variable features", subtitle = rlg, 
                                                              caption = capt) + theme_bw()
```

After talking with JMF he told me to generate two new PCAs:
- One **ommiting genes of chromosome X**
- Another **ommiting** directly the two **male samples**. 
### PCA: wo/ Males
```{r dds_NO_males}
# colData is where the metadata is stored in the 'dds' object.
colData(rlg.dds.BC1_LD.LV)
# Removal of the 2 male samples by using indexes
rlg.dds.BC1_LD.LV.noMales <- rlg.dds.BC1_LD.LV[,c(-5, -6)]
colData(rlg.dds.BC1_LD.LV.noMales)
```

```{r PCA_NO_males}
# Plot subtitle
no_males <- expression(paste(bold("rlog"), " transformation applied. Only ", bold("female "), "samples."))


# PCA
pcaData.NO_males <- plotPCA(rlg.dds.BC1_LD.LV.noMales, intgroup=c("Group"), returnData=TRUE)
percentVar.NO_males <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData.NO_males, aes(PC1, PC2, color=Group)) +
  geom_point(size=3) + geom_text_repel(aes(label = str_replace(str_replace(rownames(pcaData.NO_males), "_LV", ""),"X", ""))) +
  xlab(paste0("PC1: ",percentVar.NO_males[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar.NO_males[2],"% variance")) + 
  scale_color_manual(values = c("#2f847c", "#bdb76b")) + labs(title = "PCA of the top 500 more variable features", subtitle = no_males, caption = capt) + theme_bw()
```
### PCA: wo/ Sexual chromosomes features
```{r chromosome_information}
# Loading of the genomic features used for the read mapping: The 'gtf' file 
Tx.Sus_scrofa_11.1 <- makeTxDbFromGFF("/home/fllobet/Documents/RNA-Seq_TFM/REF/Sscrofa11.1.gtf")

# List of columns of the 'tx' object:
columns(Tx.Sus_scrofa_11.1)  
# Here I found the two interest columns: "GENEID" and "TXCHROM"

# Keys of the 'tx' object. What we can search for, I think. 
keytypes(Tx.Sus_scrofa_11.1)

# ENSEMBL IDs of all the transcripts. 
# The goal is to obtain the chromosome where the annotated features 
# of the pig genome are located. 
SS.11.IDs <- keys(Tx.Sus_scrofa_11.1, "GENEID")

# Selection of the columns "GENEID" and "TXCHROM" from all the annotated 
# features by using the 'select' function. 
SS.11.IDs_Chr <- select(Tx.Sus_scrofa_11.1, SS.11.IDs, columns = c("GENEID", "TXCHROM"), keytype = "GENEID")


# IMPORTANT: Removal of the features that were filtered intially because they
# showed a too low overall number of counts. 
SS.11.IDs_Chr_notXY <- SS.11.IDs_Chr[keep,]
# Subset of only the features located in autosomes
# Removal of chromosome X
SS.11.IDs_Chr_notXY <- SS.11.IDs_Chr_notXY[which(SS.11.IDs_Chr_notXY$TXCHROM != "X"),]
# Removal of chromosome Y
SS.11.IDs_Chr_notXY <- SS.11.IDs_Chr_notXY[which(SS.11.IDs_Chr_notXY$TXCHROM != "Y"),]
```

```{r tidying_dds_object}
# Creation of a temporary alternative 'dds' object. 
rlg.dds.BC1_LD.LV.notXY <- rlg.dds.BC1_LD.LV

# Removal of the gene names
rownames(rlg.dds.BC1_LD.LV.notXY) <- 
  str_replace(rownames(rlg.dds.BC1_LD.LV.notXY), "\\_[0-z]+", "")
keep <- 
  str_replace(keep, "\\_[0-z]+", "")

# Confirmation that the gene name has been removed
rowData(rlg.dds.BC1_LD.LV.notXY)


#SS.11.IDs_Chr_notXY.Filt
```

```{r}
# Location of the indexes of the 'dds' ojecte where the 
# autosome-located features are located
notXY.indx <- which(rownames(rlg.dds.BC1_LD.LV.notXY) %in% SS.11.IDs_Chr_notXY[,1])

rlg.dds.BC1_LD.LV.notXY <- rlg.dds.BC1_LD.LV.notXY[notXY.indx,]

dim(rlg.dds.BC1_LD.LV.notXY)
```

```{r PCA_autosomes}
# Plot subtitle
autosomes <- expression(paste(bold("rlog"), " transformation applied. Only features located within ", bold("autosomes"), "."))


# PCA
pcaData.autosomes <- plotPCA(rlg.dds.BC1_LD.LV.notXY, intgroup=c("Sex", "Group"), returnData=TRUE)
percentVar.autosomes <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData.autosomes, aes(PC1, PC2, color=Group, shape=Sex)) +
  geom_point(size=3) + geom_text_repel(aes(label = str_replace(str_replace(rownames(pcaData.autosomes), "_LV", ""),"X", ""))) +
  xlab(paste0("PC1: ",percentVar.autosomes[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar.autosomes[2],"% variance")) + 
  scale_color_manual(values = c("#2f847c", "#bdb76b")) + labs(title = "PCA of the top 500 more variable features", subtitle = autosomes, caption = capt) + theme_bw()
```

```{r removal_unsued_data}
rm(pcaData, pcaData.autosomes, rlg.dds.BC1_LD.LV.notXY, SS.11.IDs_Chr_notXY, vsd, capt, notXY.indx, percentVar, percentVar.autosomes, select, Tx.Sus_scrofa_11.1)
```

# DEG Analysis

## Definition of the reference level
In this case the DEG Analysis will be performed using **only** the **male samples**. Therefore, the design requires to only use Group as the **interest variable**. It is important to ensure that the **L** (LOW) group is considered the **reference**. 

Note that the `DESeq` function will perform the normalization itself so the `dds` object using the **raw counts** must be used as its input. 
```{r ref_level_def}
# Removal of the 2 males from the original RAW COUNTS 'dds' object using their 
# indexes:
dds.BC1_LD.LV.noMales <-dds.BC1_LD.LV[,c(-1, -3)]
# Confirmation that they have been removed:
colData(dds.BC1_LD.LV.noMales)

# Definition of the L (LOW) group as the reference group:
dds.BC1_LD.LV.noMales$Group <- relevel(dds.BC1_LD.LV.noMales$Group, 
                                           ref = "L")
```   

## DEG Analysis
Once the correct reference level has been established the **actual DEG Analysis** can be performed.
```{r DEG_Analysis}
# Update of the 'dds' object with the DEG analysis
dds.BC1_LD.LV.noMales <- DESeq(dds.BC1_LD.LV.noMales)
# Results with a == 0.1
res.noMales <- results(dds.BC1_LD.LV.noMales)
# Results with a == 0.05
res.noMales.a_0.05 <- results(dds.BC1_LD.LV.noMales, alpha=0.05)
```
In order to improve the analysis of the results several extra embellishment steps can be done.  
```{r DEGA_results_embellishment_a_0.1}
# Sorting by p.value
res.noMales.pvalueORDER <- res.noMales[order(res.noMales$pvalue),]
# Removal of NA's
res.noMales.pvalueORDER <- na.omit(res.noMales.pvalueORDER)

# Summary of the up and down regulated genes in this study:
summary(res.noMales)

# Number of significant LFCs...
sum(res.noMales$padj < 0.1, na.rm = TRUE)

# Visualization of the significant LFCs...
res.noMales[which(res.noMales$padj < 0.1),]
```
```{r DEGA_results_embellishment_a_0.05}
# Summary of the up and down regulated genes in this study:
summary(res.noMales.a_0.05)

# Number of significant LFCs...
sum(res.noMales.a_0.05$padj < 0.05, na.rm = TRUE)

# Visualization of the significant LFCs...
res.noMales.a_0.05[which(res.noMales.a_0.05$padj < 0.05),]

```

```{r results_writing}
# Storing the results inot a CSV file
write.table(as.data.frame(res.noMales.pvalueORDER), 
          file="BC1_LD.LV.DEGanalysis_noMales_results.csv", sep = ";")
```

### MA Plot
A good option to visualize the RNA-Seq results is to generate a MA Plot. 
```{r}
# MA plot
plotMA(res.noMales)
```
Another interesting plot is the **QQ-plot**:
```{r}
# Function form the package 'snpStats'
qq.chisq(-2*log(res.noMales$pvalue),df=2,pvals=T,overdisp=T, 
         col = "#4d5d53", col.shade = "#ace1af", main = "QQ plot of p-values.\nNo sex effect: only female samples")
```

## Addition of LFC thershold

A part from the **p.adjusted value** we use an extra criterion:
**Fold-Change (FC)**. We use 2 different LFC thresholds: 
- Absolute 1.2 of FC 
- Absolute 1.5 of FC 
Which in turn end up creating 4 combinations: 
- 1.2 of FC + alpha 0.1 
- 1.5 of FC + alpha 0.05
- 1.2 of FC + alpha 0.1 
- 1.5 of FC + alpha 0.05

### Alpha = 0.1

```{r FC_1.2+a_0.1}
# FC = 1.2
lg2_fc1.2 <- log2(1.2)

# Selection of the features with a log2 FC higher or lower than 1.2:
# Higher
res.BC1_PI.LV_above_fc.12 <- res.BC1_PI.LV[which(res.BC1_PI.LV$log2FoldChange >= lg2_fc1.2),]
# Lower 
res.BC1_PI.LV_below_fc.12 <- res.BC1_PI.LV[which(res.BC1_PI.LV$log2FoldChange <= -lg2_fc1.2),]
# Merge
res.BC1_PI.LV_abv.blw_fc.12 <- rbind(res.BC1_PI.LV_above_fc.12, res.BC1_PI.LV_below_fc.12)

write.table((rownames(res.BC1_PI.LV.a_0.05[which(res.BC1_PI.LV.a_0.05$padj < 0.05),])),"./toBIOstars.sij_005.csv", sep = ";")

```

The number of significant genes gets reduced from `r sum(res.BC1_PI.LV$padj < 0.1, na.rm = TRUE)` to `r sum(res.BC1_PI.LV_abv.blw_fc.12$padj < 0.1, na.rm = TRUE)`.

Number of up-regulated genes: `r sum(res.BC1_PI.LV_above_fc.12$padj < 0.1, na.rm = TRUE)`
Number of down-regulated genes: `r sum(res.BC1_PI.LV_below_fc.12$padj < 0.1, na.rm = TRUE)`
```{r FC_1.5+a_0.1}
# FC = 1.5
lg2_fc1.5 <- log2(1.5)

# Selection of the features with a log2 FC higher or lower than 1.5:
# Higher
res.BC1_PI.LV_above_fc.15 <- res.BC1_PI.LV[which(res.BC1_PI.LV$log2FoldChange >= lg2_fc1.5),]
# Lower 
res.BC1_PI.LV_below_fc.15 <- res.BC1_PI.LV[which(res.BC1_PI.LV$log2FoldChange <= -lg2_fc1.5),]
# Merge
res.BC1_PI.LV_abv.blw_fc.15 <- rbind(res.BC1_PI.LV_above_fc.15, res.BC1_PI.LV_below_fc.15)
```

The number of significant genes get reduced from `r sum(res.BC1_PI.LV$padj < 0.1, na.rm = TRUE)` to `r sum(res.BC1_PI.LV_abv.blw_fc.15$padj < 0.1, na.rm = TRUE)`.

Number of up-regulated genes: `r sum(res.BC1_PI.LV_above_fc.15$padj < 0.1, na.rm = TRUE)`
Number of down-regulated genes: `r sum(res.BC1_PI.LV_below_fc.15$padj < 0.1, na.rm = TRUE)`

### Alpha = 0.05

```{r FC_1.2+a_0.1}
# FC = 1.2
lg2_fc1.2 <- log2(1.2)

# Selection of the features with a log2 FC higher or lower than 1.2:
# Higher
res.BC1_PI.LV_05above_fc.12 <- res.BC1_PI.LV.a_0.05[which(res.BC1_PI.LV.a_0.05$log2FoldChange >= lg2_fc1.2),]
# Lower 
res.BC1_PI.LV_05below_fc.12 <- res.BC1_PI.LV.a_0.05[which(res.BC1_PI.LV.a_0.05$log2FoldChange <= -lg2_fc1.2),]
# Merge
res.BC1_PI.LV_05abv.blw_fc.12 <- rbind(res.BC1_PI.LV_05above_fc.12, res.BC1_PI.LV_05below_fc.12)
```

The number of significant genes gets reduced from `r sum(res.BC1_PI.LV.a_0.05$padj < 0.05, na.rm = TRUE)` to `r sum(res.BC1_PI.LV_05abv.blw_fc.12$padj < 0.05, na.rm = TRUE)`.

Number of up-regulated genes: `r sum(res.BC1_PI.LV_05above_fc.12$padj < 0.05, na.rm = TRUE)`
Number of down-regulated genes: `r sum(res.BC1_PI.LV_05below_fc.12$padj < 0.05, na.rm = TRUE)`

```{r FC_1.5+a_0.05}
# FC = 1.5
lg2_fc1.5 <- log2(1.5)

# Selection of the features with a log2 FC higher or lower than 1.5:
# Higher
res.BC1_PI.LV_05above_fc.15 <- res.BC1_PI.LV.a_0.05[which(res.BC1_PI.LV.a_0.05$log2FoldChange >= lg2_fc1.5),]
# Lower 
res.BC1_PI.LV_05below_fc.15 <- res.BC1_PI.LV.a_0.05[which(res.BC1_PI.LV.a_0.05$log2FoldChange <= -lg2_fc1.5),]
# Merge
res.BC1_PI.LV_05abv.blw_fc.15 <- rbind(res.BC1_PI.LV_05above_fc.15, res.BC1_PI.LV_05below_fc.15)

write.table((rownames(res.BC1_PI.LV_05abv.blw_fc.15[which(res.BC1_PI.LV_05abv.blw_fc.15$padj < 0.05),])),"./toBIOstars.sij_1.5_005.csv", sep = ";")
```

The number of significant genes gets reduced from `r sum(res.BC1_PI.LV.a_0.05$padj < 0.05, na.rm = TRUE)` to `r sum(res.BC1_PI.LV_05abv.blw_fc.15$padj < 0.05, na.rm = TRUE)`.

Number of up-regulated genes: `r sum(res.BC1_PI.LV_05above_fc.15$padj < 0.05, na.rm = TRUE)`
Number of down-regulated genes: `r sum(res.BC1_PI.LV_05below_fc.15$padj < 0.05, na.rm = TRUE)`
